pkg/agent/route/route_linux.go

```diff
@@ -836,51 +836,74 @@ func (c *Client) AddRoutes(podCIDR *net.IPNet, nodeName string, nodeIP, nodeGwIP
+	}
 	// If WireGuard is enabled, create a route via WireGuard device regardless of the traffic encapsulation modes.
 	if c.networkConfig.TrafficEncryptionMode == config.TrafficEncryptionModeWireGuard {
-		route.LinkIndex = c.nodeConfig.WireGuardConfig.LinkIndex
-		route.Scope = netlink.SCOPE_LINK
+		podCIDRRoute.LinkIndex = c.nodeConfig.WireGuardConfig.LinkIndex
+		podCIDRRoute.Scope = netlink.SCOPE_LINK
 		if podCIDR.IP.To4() != nil {
-			route.Src = c.nodeConfig.GatewayConfig.IPv4
+			podCIDRRoute.Src = c.nodeConfig.GatewayConfig.IPv4
 		} else {
-			route.Src = c.nodeConfig.GatewayConfig.IPv6
+			podCIDRRoute.Src = c.nodeConfig.GatewayConfig.IPv6
 		}
+		routes = append(routes, podCIDRRoute)
 	} else if c.networkConfig.NeedsTunnelToPeer(nodeIP, nodeTransportIPAddr) {
 		if podCIDR.IP.To4() == nil {
 			// "on-link" is not identified in IPv6 route entries, so split the configuration into 2 entries.
-			routes = []*netlink.Route{
-				{
-					Dst:       &net.IPNet{IP: nodeGwIP, Mask: net.IPMask{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}},
-					LinkIndex: c.nodeConfig.GatewayConfig.LinkIndex,
-				},
-			}
+			// TODO: Kernel >= 4.16 supports adding IPv6 route with onlink flag. Delete this route after Kernel version
+			//       requirement bump in future.
+			routes = append(routes, &netlink.Route{
+				Dst:       nodeGwIPNetv6,
+				LinkIndex: c.nodeConfig.GatewayConfig.LinkIndex,
+			})
 		} else {
-			route.Flags = int(netlink.FLAG_ONLINK)
+			podCIDRRoute.Flags = int(netlink.FLAG_ONLINK)
+			podCIDRRoute.LinkIndex = c.nodeConfig.GatewayConfig.LinkIndex
+			podCIDRRoute.Gw = nodeGwIP
+			routes = append(routes, podCIDRRoute)
 		}
-		route.LinkIndex = c.nodeConfig.GatewayConfig.LinkIndex
-		route.Gw = nodeGwIP
 	} else if c.networkConfig.NeedsDirectRoutingToPeer(nodeIP, nodeTransportIPAddr) {
 		// NoEncap traffic to Node on the same subnet.
 		// Set the peerNodeIP as next hop.
-		route.Gw = nodeIP
+		podCIDRRoute.Gw = nodeIP
+		routes = append(routes, podCIDRRoute)
 	} else {
 		// NetworkPolicyOnly mode or NoEncap traffic to a Node on a different subnet.
 		// Routing should be handled by a route which is already present on the host.
-		return nil
+		klog.InfoS("Skip adding routes to peer", "node", nodeName, "ip", nodeIP, "podCIDR", podCIDR)
 	}
-	routes = append(routes, route)
-
+	routeToNodeGwIPv6Replaced := false
+	routeToPodCIDRReplaced := false
 	for _, route := range routes {
 		if err := netlink.RouteReplace(route); err != nil {
 			return fmt.Errorf("failed to install route to peer %s (%s) with netlink. Route config: %s. Error: %v", nodeName, nodeIP, route.String(), err)
 		}
+		if route.Dst == nodeGwIPNetv6 {
+			routeToNodeGwIPv6Replaced = true
+		}
+		if route.Dst == podCIDR {
+			routeToPodCIDRReplaced = true
+		}
 	}
-
-	if podCIDR.IP.To4() == nil {
+	if !routeToNodeGwIPv6Replaced && utilnet.IsIPv6(nodeGwIP) {
```

Comment:
Not only routeToNodeGwIPNetv6, we also need to remove routeToPodCIDRReplaced when it's NoEncap mode in different subnet. And second thought about how to identify stale routes:
```
staleRouteDsts := sets.NewString(podCIDR.String())
if utilnet.IsIPv6(nodeGwIP) {
	staleRouteDsts.Insert(&net.IPNet{IP: nodeGwIP, Mask: net.CIDRMask(128, 128).String())
}
if c.networkConfig.TrafficEncryptionMode == config.TrafficEncryptionModeWireGuard {
	route.LinkIndex = c.nodeConfig.WireGuardConfig.LinkIndex
	route.Scope = netlink.SCOPE_LINK
	if podCIDR.IP.To4() != nil {
		route.Src = c.nodeConfig.GatewayConfig.IPv4
	} else {
		route.Src = c.nodeConfig.GatewayConfig.IPv6
	}
	staleRouteDsts.Delete(podCIDR.String())
} else if ... {
...
}
...
for dst := range staleRouteDsts {
        if err := netlink.RouteDel(&netlink.Route{Dst: dst}); err != nil {
	...
}
```

Reply:
I am unsure about the CIDR route for Pods in NoEncap mode & different subnets. Do you think we may delete the route added by users or other CNIs mistakenly if we only use the Pod CIDR as dst IP to identify a route?